File: JVM/app/src/main/java/mntone/univschedule/demo/core/UnivScheduleAppContext.java
package mntone.univschedule.demo.core;

import org.jdeferred.Promise;
import org.jdeferred.android.AndroidDeferredManager;

import java.util.HashMap;

import mntone.univschedule.core.*;
import mntone.univschedule.client.UniversityScheduleClient;

public final class UnivScheduleAppContext
{
	private static UnivScheduleAppContext _context;

	public static UnivScheduleAppContext getContext()
	{
		if( _context == null )
		{
			_context = new UnivScheduleAppContext();
		}
		return _context;
	}

	private final AndroidDeferredManager _deferredManager;
	private final UniversityScheduleClient _client;

	private UniversitiesResponse _universities = null;
	private HashMap<String, ClassesResponse> _classesMap = null;

	public UnivScheduleAppContext()
	{
		this._deferredManager = new AndroidDeferredManager();
		this._client = new UniversityScheduleClient( "testing123" );

		this._classesMap = new HashMap<>();
	}

	public Promise<UniversitiesResponse, Throwable, Void> getUniversitiesAsync()
	{
		return this._deferredManager.when(
			() ->
			{
				if( this._universities == null )
				{
					this._universities = this._client.getUniversities();
					return this._universities;
				}
				return this._universities;
			} );
	}

	public Promise<ClassesResponse, Throwable, Void> getClassesAsync( final String universityScreenName )
	{
		return this._deferredManager.when(
			() ->
			{
				if( !this._classesMap.containsKey( universityScreenName ) )
				{
					final ClassesResponse classes = this._client.getClasses( universityScreenName );
					this._classesMap.put( universityScreenName, classes );
					return classes;
				}

				return this._classesMap.get( universityScreenName );
			} );
	}
}

File: NET/UniversitySchedule.Client.Demo/Core/UnivScheduleAppContext.cs
using Mntone.UniversitySchedule.Core;
using Mntone.UniversitySchedule.Client;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Mntone.UniversitySchedule.Client.Demo.Core
{
	public sealed class UnivScheduleAppContext
	{
		public static UnivScheduleAppContext Context
		{
			get { return _Context ?? ( _Context = new UnivScheduleAppContext() ); }
		}
		private static UnivScheduleAppContext _Context;


		private readonly UniversityScheduleClient _client;

		private UniversitiesResponse _universities;
		private Dictionary<string, ClassesResponse> _classesMap = new Dictionary<string, ClassesResponse>();

		private UnivScheduleAppContext()
		{
			this._client = new UniversityScheduleClient( "testing123" );
		}

		public Task<UniversitiesResponse> GetUniverstiesAsync()
		{
			if( this._universities != null )
			{
				return Task.FromResult( this._universities );
			}
			return this._client.GetUniversitiesAsync();
		}

		public Task<ClassesResponse> GetClassesAsync( string screenName )
		{
			if( this._classesMap.ContainsKey( screenName ) )
			{
				return Task.FromResult( this._classesMap[screenName] );
			}
			return this._client.GetClassesAsync( screenName );
		}
	}
}